### PR TITLE
 feat(hooks): add useHitsPerPage

### PR DESCRIPTION
**Pull request by tkrugg**
**Summary**

 adds useHierarchicalMenu

**Result**

closes https://algolia.atlassian.net/browse/FX-446

### CODE REVIEW

File: examples/hooks/App.css [hunk 1]
```diff
@@ -31,7 +31,7 @@ body {
 .Search-header {
   display: grid;
   align-items: flex-start;
-  grid-template-columns: 1fr auto;
+  grid-template-columns: 1fr auto auto;
```
Comment (tkrugg) [inline examples/hooks/App.css:34]: this changes the style slightly so we can fit in the HitsPerPage widge

Screenshot 2021-11-22 at 16 37 44
t

File: examples/hooks/index.html [hunk 2]
```diff
@@ -15,6 +15,9 @@
       rel="stylesheet"
       href="https://cdn.jsdelivr.net/npm/instantsearch.css@7/themes/satellite-min.css"
     />
+    <script>
+      window.__DEV__ =  true
+    </script>
```
Comment (tkrugg) [inline examples/hooks/index.html:20]:
```suggestion
```

File: examples/hooks/index.tsx [hunk 3]
```diff
@@ -4,3 +4,7 @@ import ReactDOM from 'react-dom';
 import { App } from './App';
 
 ReactDOM.render(<App />, document.getElementById('root'));
+
+if (module.hot) {
+  module.hot.accept();
+}
```
Comment (tkrugg) [inline examples/hooks/index.tsx:10]:
```suggestion
```

File: examples/hooks/components/HitsPerPage.tsx [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+export type HitsPerPageProps = React.ComponentProps<'div'> &
+  UseHitsPerPageProps;
+
+export function HitsPerPage(props: HitsPerPageProps) {
+  const { items, refine } = useHitsPerPage(props);
+  const { value: currentValue } =
+    items.find(({ isRefined }) => isRefined) || {};
+
+  const handleChange = useCallback(
```
Comment (Haroenv) [inline examples/hooks/components/HitsPerPage.tsx:13]: is this required? the refine callback should be static already no?

File: packages/react-instantsearch-hooks/README.md [hunk 5]
```diff
@@ -620,7 +620,7 @@ function RefinementList(props) {
 
 ### `useSortBy`
 
-> `(props: UseSortByProps) => SortByRenderState`
+> `(props: UseSortByProps) => HitsPerPageRenderState`
```
Comment (Haroenv) [inline packages/react-instantsearch-hooks/README.md:623]:
```suggestion
> `(props: UseSortByProps) => SortByRenderState `
```

File: packages/react-instantsearch-hooks/src/useHitsPerPage.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+import { useConnector } from './useConnector';
+
+import type {
+  HitsPerPageConnectorParams,
+  HitsPerPageWidgetDescription,
+} from 'instantsearch.js/es/connectors/hits-per-page/connectHitsPerPage';
+
+export type UseHitsPerPageProps = HitsPerPageConnectorParams;
+
+export function useHitsPerPage(props?: UseHitsPerPageProps) {
```
Comment (Haroenv) [inline packages/react-instantsearch-hooks/src/useHitsPerPage.ts:12]:
```suggestion
export function useHitsPerPage(props: UseHitsPerPageProps) {
```

**Review by Haroenv — COMMENTED** [review]
👌

File: examples/hooks/components/HitsPerPage.tsx (hunk 4 above)
Comment (francoischalifour) [reply examples/hooks/components/HitsPerPage.tsx:13]: Indeed, we don't need this overhead. Watching the reference is more costly than creating a new function reference here.

File: examples/hooks/components/HitsPerPage.tsx [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+      refine(event.target.value);
+    },
+    [refine]
+  );
+  return (
+    <div className={cx('ais-HitsPerPage', props.className)}>
+      <select
+        className="ais-HitsPerPage-select"
+        onChange={handleChange}
+        value={`${currentValue}`}
```
Comment (francoischalifour) [inline examples/hooks/components/HitsPerPage.tsx:24]: This better communicates the intent?

```suggestion
        value={String(currentValue)}
```

File: examples/hooks/components/HitsPerPage.tsx [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+      >
+        {items.map((item) => (
+          <option
+            key={item.label + item.value}
```
Comment (francoischalifour) [inline examples/hooks/components/HitsPerPage.tsx:28]: Why do we need the label in the key here? React keys are scoped to their `map`.

File: packages/react-instantsearch-hooks/src/__tests__/useHitsPerPage.test.tsx [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+          isRefined: false,
+          label: '8 hits per page',
+          value: 8,
+        },
+      ],
+    });
+
+    await waitForNextUpdate();
+
+    // Initial render state from manual `getWidgetRenderState`
```
Comment (francoischalifour) [inline packages/react-instantsearch-hooks/src/__tests__/useHitsPerPage.test.tsx:44]:
```suggestion
    // InstantSearch.js state from the `render` lifecycle step
```

File: examples/hooks/components/HitsPerPage.tsx (hunk 8 above)
Comment (tkrugg) [reply examples/hooks/components/HitsPerPage.tsx:28]: i picked this up looking at [the Selector component from InstantSearch.js](https://github.com/algolia/instantsearch.js/blob/master/src/components/Selector/Selector.tsx#L39). 
I think it's possible to have twice the same value in items (though the isn't a use case for is here).
If we agree we can remove it I guess.

File: examples/hooks/components/HitsPerPage.tsx (hunk 7 above)
Comment (tkrugg) [reply examples/hooks/components/HitsPerPage.tsx:24]: works for me!

File: examples/hooks/components/HitsPerPage.tsx (hunk 8 above)
Comment (tkrugg) [reply examples/hooks/components/HitsPerPage.tsx]: removed in https://github.com/algolia/react-instantsearch/pull/3205/commits/ebb0c124098660e61bd1293d1d145f93ed7a94dc

File: examples/hooks/components/HitsPerPage.tsx (hunk 4 above)
Comment (tkrugg) [reply examples/hooks/components/HitsPerPage.tsx]: removed in https://github.com/algolia/react-instantsearch/pull/3205/commits/ebb0c124098660e61bd1293d1d145f93ed7a94dc

File: packages/react-instantsearch-hooks/README.md [hunk 10]
```diff
@@ -620,7 +620,7 @@ function RefinementList(props) {
 
 ### `useSortBy`
 
-> `(props: UseSortByProps) => SortByRenderState`
+> `(props: UseSortByProps) => SortByRenderState `
```
Comment (tkrugg) [inline packages/react-instantsearch-hooks/README.md:623]:
```suggestion
> `(props: UseSortByProps) => SortByRenderState`
```

**Review by francoischalifour — APPROVED** [review]

**Review by Haroenv — APPROVED** [review]

**Pull request merged**